### DISCUSSION TITLE
Static / SSR page crashes on load with `Uncaught Error: Minified React error #423;`

### DISCUSSION THREAD

**Discussion by kavitang** [vercel/next.js#51825] [Help]
### Summary

In our Next 13 app app router, a previously working route is now suddenly crashing with a very hard to debug issue.

```
Uncaught Error: Minified React error #423; visit https://reactjs.org/docs/error-decoder.html?invariant=423 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
```

It does not point to any helpful files or sources and is not being thrown in development mode (no warning either). I have searched for possible ways to debug this, but none have yielded results so far. The actual file linked in the hydration is the Sentry component, so again it's a dead end. 

In addition, this error is also followed by a hydration error

```
TypeError: Cannot read properties of undefined (reading 'call')
 at h (webpack-78bb0dd2075c1e83.js:1:173)
    at 5683 (46450.6c4228780f1915f4.js:1:3121)
    at h (webpack-78bb0dd2075c1e83.js:1:173)
    at x (40749-b3ad58cb3bb1f3b1.js:26:4272)
    at E (40749-b3ad58cb3bb1f3b1.js:26:2856)
    at iv (bce60fc1-c136354741099a55.js:9:112270)
    at oR (bce60fc1-c136354741099a55.js:9:90087)
    at bce60fc1-c136354741099a55.js:9:86345
    at bce60fc1-c136354741099a55.js:9:86352
    at ok (bce60fc1-c136354741099a55.js:9:86457)
```
I checked `46450.6c4228780f1915f4.js:1:3121` chunk and error points to an import 

```
var s = l(9268)
          , n = l.n(i)
          , d = l(76394)
          , o = l.n(d)
          , c = l(16106)
          , u = l(92678)   <------- This one
          , x = l(16933)
```
After searching for this value I found that this import is actually a @iconscount/react-unicons icon `uil-angle-left`. I commented this out, and another import then started throwing this error, which is hook that reads from local storage. All of this was working previously, and has been correctly labelled as client components where needed. 

After spending countless hours on this, I'm yet to find the cause for the crash and am not sure how to continue debugging this more effectively.

I also checked for potential issues mentioned here https://github.com/vercel/next.js/issues/50245, no luck.

Any explanations on what else could be a cause for this issue would be extremely helpful, or even how I can go about debugging this.

Thanks!

**Comment by icyJoseph** [comment]
>  I commented this out, and another import then started throwing this error, which is hook that reads from local storage. All of this was working previously, and has been correctly labelled as client components where needed.

Could it be some unexpected value in the local storage? I the component is a client component, it'll still run server side, just so you know. So that might just cause the component to throw. 

Is there any repo/live site we could take a look at?

**Reply by kavitang** [reply]
@icyJoseph - I checked that case in incognito, we use `usehooks-ts` `useLocalStorage` hook for this, which handles SSR, and there is no value in local storage to affect the UI. 

Also before the local storage hook, it's the `iconscout/react-unicons` import that seems to be causing an issue. The live site for this is [here](http://buildappeal.dev/gallery), this route is statically served (the home route works and if you navigate to gallery from there it works fine as well). I tried using `suppressHydrationWarning` to see if that adds a temporary fix, but that only worked in my local environment and in a preview deployment. The production deployment continues to fail with the same error.

**Reply by icyJoseph** [reply]
Mmm alright, let me think a bit further. At least we have discarded a few things.

**Reply by kavitang** [reply]
@icyJoseph Thank you! I testing converting all `dynamic` imports to standard imports to see if that makes a difference, will update here once I've tried a few more things.

**Reply by icyJoseph** [reply]
Yeah it is definitely a lazy loaded component that's crashing out here.

![Screenshot 2023-06-26 at 22 34 54](https://github.com/vercel/next.js/assets/21013447/c4d3ae3d-d8f9-4009-b8a3-5a1bc3572a95)

**Reply by icyJoseph** [reply]
A little more data:

![Screenshot 2023-06-26 at 22 35 43](https://github.com/vercel/next.js/assets/21013447/41d84edb-431e-4bb9-8581-9a856f77bda0)

**Reply by kavitang** [reply]
@icyJoseph - Thank you, could you share which file you added the breakpoint to? Is it one of the files in the error trace? It would be very helpful to know how to get this information in such cases

**Reply by icyJoseph** [reply]
On it! I just down want to throw you off - just double checking that I am after the core issue.

**Reply by icyJoseph** [reply]
Alright, when the error comes, click this one in the React summary: bce60fc1-c136354741099a55.js:9:113116

Place a debugger, and reload, from there you ought to step up to the stack (arrow up)

Then it is all about your app knowledge, but this prop looks suspicious:

```
{
    "id": "All",
    "title": "All",
    "value": "all",
    "icon": "https://qhjjmjbamculrcdfmnho.supabase.co/storage/v1/object/public/files/spaceIcons/all.svg"
}
```

Maybe it is ok, IDK - 

It looks like file 40749-6b3381fcde7e4ebb.js might have the key to the mystery?

**Reply by icyJoseph** [reply]
Or perhaps is it? static/chunks/46450.4e9bfe3eb3efc9a7.js, see now I am producing more noise than help, but it feels like we are closing on the issue

**Reply by icyJoseph** [reply]
Also found this one at top of the stack, which is the root issue: https://legacy.reactjs.org/docs/error-decoder.html/?invariant=418

Something from the server changing on the client...

**Reply by kavitang** [reply]
@icyJoseph - I removed all the dynamic imports in this route, but it does not seem to have resolved the issue. The error remains the exact same. I'll try and work through the stack trace to see if I can spot more information about the lazy loaded component.. 46450 is the culprit for sure, it has the missing import (the component itself has no direct dynamic imports, neither do it's children)... 40749 is the Sentry file, I think that is being pulled into the stack due to the error... thank you for sharing the steps, I'll post back with updates once I run through this 🙏

**Reply by kavitang** [reply]
@icyJoseph - An update, I have not been able to identify the lazy loaded component. All dynamic imports have been removed but the error persists. The missing import continues to point to the `uil-angle-left` component, which is not being lazy loaded. I removed this import, and it then switches on to more missing imports. 

Is there an effective way for me to debug the Hydration error? There are no warnings in the development environment, this error is only happening with the production app.

**Reply by icyJoseph** [reply]
Mmm can you trigger it with a local build? 

Wonder if there's some kind of problem with the size of the app? Where is it hosted?

Do you have middleware activated?

**Reply by kavitang** [reply]
Yes I can trigger it with a local build, just not in the development mode. App has no middleware and is hosted on Vercel. I'm trying to comment out parts of the component that's throwing this error to see if I can pinpoint the exact source of the hydration error. Are there any additional commands/settings I could use in the local build that might give me more information?

**Reply by icyJoseph** [reply]
Alright, but if you can trigger it locally using the production build, we can discard locale issues.

Weird that no request fails at all, right? Why else would a lazy component fail. 

What is the library that loads these icons? I wonder if you could also place an error boundary around the affected area. Perhaps we can prevent a total app breakage (not sure this issue can be captured by those)

I wonder if there's somewhere a process.env.NODE_ENV condition, that's not being taken into consideration. 

I checked the HTML from the server, and rendered by the browser without JavaScript, and they seem identical...

**Reply by kavitang** [reply]
Yes, I also compared the source and the browser HTML, no noticeable difference. But I was able to narrow down the issue to three specific issues/imports. 

1. The `@iconscout/react-unicons` import for `uil-angle-left` and right. This import fails and the component that uses it is setting the nav buttons visibility based on next / previous items.
2. The `useLocalStorage` hook import
3. A `Modal` component that is importing a `@headlessui/react` Dialog component. This component also has a Share component within it which is also using `@iconscout/react-unicons` `uil-angle-left`. 

I commented all three and related code out and the page loads perfectly fine in the local build. The one thing I noticed is that in the `page-12345.js` chunk for `/gallery` these imports are all present as functions, but the component within the page does not seem to be recognising them(I'm not entirely clear on how this works so my assumption is that either this should work or something is being built incorrectly). Also you are right no requests are failing... 

I'm not exactly sure how to replace these, but for now I'm trying to move these out to separate client components, and add all visibility etc behind useEffects.  

I also moved the useLocalStorage to a custom hook that is using `window.localStorage` directly, with a `window` check and default returns so I know it won't crash on the server. But I'm not sure why the imports are failing and what I can do to fix this.. There is nothing special about these imports, and we use the `@iconscout/react-unicons` library in other places successfully..

<img width="888" alt="Screenshot 2023-06-27 at 9 25 23 AM" src="https://github.com/vercel/next.js/assets/96101675/083a2b92-2a4f-4f0b-9808-b7ecda4b3b7b">

**Reply by icyJoseph** [reply]
Do you experience the same if you import directly from the `@iconscout/react-unicons/ui-angle-left` path?

**Reply by kavitang** [reply]
All our imports are direct module imports `import UilAngleLeft from "@iconscout/react-unicons/icons/uil-angle-left";`

**Reply by icyJoseph** [reply]
uff what a hard nut to crack 😅 

Could it be something with the route groups?

**Reply by kavitang** [reply]
@icyJoseph - This has been resolved. It was the spaces navigations button (one of which was hidden on load) that was causing the hydration issue. I moved that check behind a useEffect and all errors have been resolved. I'm not sure why the imports were throwing the `call` error if it was only a hydration error, wasted a lot of time trying to follow that one 😅 . But it's working fine now, thanks for all your help!

**Reply by icyJoseph** [reply]
No worries! Don't forget to mark as resolved and close the issue :)

**Reply by suvasishm** [reply]
> This has been resolved. It was the spaces navigations button (one of which was hidden on load) that was causing the hydration issue. I moved that check behind a useEffect and all errors have been resolved. 

@kavitang may I have some more information on the fix you made to solve the issue? I am in a similar situation - the production build is throwing the error but no error on local/dev unminified build.

**Reply by kavitang** [reply]
@suvasishm - If it's a hydration error, then running a check to see if any components return elements on load that may be different from the server content will help. Anything loaded after a userEffect is clear of the hydration checks, it's your default returns  (like in my case the navigation had a hidden left button)(looking further into hydration checks will also help). I would recommend comparing your page source (Right click > View Page Source) with your browser HTML (Elements) to see which components should not be present on the server but still show up. This may not be very helpful sorry, but it's very app specific and requires some tedious walking through your document tree on the server and client.

**Reply by suvasishm** [reply]
@kavitang - Thank you for your response. Both the loader and page source vs element check pointers are going to help. We have loaders on these pages.
